### PR TITLE
ID Fields

### LINKED ISSUE #46: ID field
In order to improve metadata about which fields represent the IDs of an entity it would be possible to declare as field like:

```js
const Customer = 
    entity('Customer', {
        id: id(Number),    <---- id definition
        name: field(String),
        isVIP() {
            ...
        }
    })
```

for composite IDs (ex: id + SSN):

```js
const Customer = 
    entity('Customer', {
        id: id(Number),    <---- id definition
        SSN: id(Number),    <---- id definition
        name: field(String),
        isVIP() {
            ...
        }
    })
```

metadata:

```js
Customer.prototype.meta.schema.id.isID //true
Customer.prototype.meta.schema.SSN.isID //true
```

This would help in:
- CLI: better use case generation
- 2Knex: better repository generation (no ID info required)
- 2REST and 2GQL: better resolver code

### PR DESCRIPTION
**Pull request by thgpdr**
Fixes #46

## Proposed Changes

Adding ID field option on metadata. 

Took the liberty to add a type validation for ID fields (only ```Number``` and ```String```), if the wrong type is sent, gotu will ignore the isID option. Hope it's consensus among all that we need something like that, but I did not implemented nothing fancy for now.

Any feedbacks?
Thanks 

### Author/Contributor
- [x] If documentation is needed for this change, has that been included in this pull request
- [x] Remember to check if code coverage decrease, if so, please implement the necessary tests

### Reviewing Maintainer
- [ ] Label as `breaking` if this is a large fundamental change
- [ ] Label as either `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`, or `performance`

### CODE REVIEW

File: src/field.js [hunk 1]
```diff
@@ -1,14 +1,18 @@
 class Field {
   constructor(type, options = {}) {
-    this.name = ""
+    this.name = ''
     this.type = type
     this.options = options
+    this.isId = !!options.isId
     this._validations = null
+
+    if (this.isId)
```
Comment (italojs) [inline src/field.js]: why not
``` js
if ((this.isId) && (!(this.type === Number) && !(this.type === String))) this.isId = false

File: test/entity/multipleFieldsEntity.js [hunk 2]
```diff
@@ -1,135 +1,163 @@
...
+      assert.strictEqual(instance['field4'], true)
+      assert.deepStrictEqual(instance['field5'], newEntity)
+      assert.deepStrictEqual(instance['field5'].f1, 'abc')
+      assert.deepStrictEqual(instance['field6'][0], newEntity)
+      assert.deepStrictEqual(instance['field6'][0].f1, 'abc')
+    })
+
+    it('should have multiple instances with isolated valued from each other', () => {
+      //given
+      const instance1 = givenAnEntityWithMultipleFields()
```
Comment (italojs) [inline test/entity/multipleFieldsEntity.js]: why not use an array?
``` js
const instances = Promise.all(Array(3).map(x => givenAnEntityWithMultipleFields()))
instances[0]
instances[1]
instances[2]

**Review by italojs — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -216,6 +216,30 @@ const User =
     })
 ```
 
+### ID Fields
+
+It is possible to declare a field as an id, wich will take impact on the entity metadata and our glue libs.
```
Comment (dalssoft) [inline README.md]: "It is possible to declare a field as an ID. This metadata will be used by glues to enrich the infrastructure interfaces (Database, REST, GraphQL, etc)."

File: src/field.js [hunk 4]
```diff
@@ -1,14 +1,18 @@
...
+      if (!(this.type === Number) && !(this.type === String)) this.isId = false
```
Comment (dalssoft) [inline src/field.js]: I don't think it should have any constrain for the type of the ID at domain level. Let the infrastructure glue or layer decides if such restriction should be applied.

File: test/entity/multipleFieldsEntity.js [hunk 5]
```diff
@@ -1,135 +1,163 @@
 const { entity } = require('../../src/entity')
 const { field } = require('../../src/field')
 const assert = require('assert')
+const { id } = require('../../src/id')
 
 describe('An entity', () => {
+  describe('with multiple fields', () => {
```
Comment (dalssoft) [inline test/entity/multipleFieldsEntity.js]: for some reason it looks like a whole new test session was created in this PR, but looking at the merge, it looks like only a few lines were added or changed. the ideal is to try to keep the PRs clean, showing only what has changed.

File: test/entity/multipleFieldsEntity.js [hunk 6]
```diff
@@ -1,135 +1,163 @@
...
+    }
+
+    it('should initiate', () => {
+      //given
+      const instance = givenAnEntityWithMultipleFields()
+      //then
+      assert.equal(instance.meta.name, 'An entity')
+    })
+
+    it('should set a value to multiple fields', () => {
```
Comment (dalssoft) [inline test/entity/multipleFieldsEntity.js]: ID fields have been added to the entity to be tested, but the values and behaviors are not being tested in all scenarios (eg `field7` to `field11`)

File: src/test.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+const { entity } = require('./entity')
```
Comment (dalssoft) [inline src/test.js]: what is this file / test for?

**Review by dalssoft — CHANGES_REQUESTED** [review]

File: test/entity/multipleFieldsEntity.js (hunk 5 above)
Comment (thgpdr) [reply test/entity/multipleFieldsEntity.js]: Oh my.. My prettier just changed the hole file, let me fix it.

File: test/entity/multipleFieldsEntity.js (hunk 2 above)
Comment (thgpdr) [reply test/entity/multipleFieldsEntity.js]: Sorry, my linter changed the hole file.. This test was not implemented by this pr, but hey, I can fix it on another pr, no problem

File: src/test.js (hunk 7 above)
Comment (thgpdr) [reply src/test.js]: file removed

File: test/entity/multipleFieldsEntity.js (hunk 6 above)
Comment (thgpdr) [reply test/entity/multipleFieldsEntity.js]: I changed the tests, can you give me a new feedback on those?
thanks @dalssoft

**Comment by dalssoft** [comment]
I think we are going in the right direction. One point: is the PR name using the correct semantics? https://github.com/herbsjs/gotu/blob/master/.github/CONTRIBUTING.md

**Comment by dalssoft** [comment]
Something to think about before we put the idea of IDs in the lib: What should the behavior of `.validate()` be?
Ex: If a field is an ID should it allow null (`allowNull: false`)?

**Comment by jhomarolo** [comment]
> Something to think about before we put the idea of IDs in the lib: What should the behavior of `.validate()` be?
> Ex: If a field is an ID should it allow null (`allowNull: false`)?

For me it's not a problem that the ID is blank/null, because sometimes, in an insertion case, the ID only exists after an insertion in the database.

**Comment by dalssoft** [comment]
> For me it's not a problem that the ID is blank/null, because sometimes, in an insertion case, the ID only exists after an insertion in the database.
Good point. So what should be the DX here?

== scenario 1 - 
`.validate()`

**Comment by dalssoft** [comment]
> For me it's not a problem that the ID is blank/null, because sometimes, in an insertion case, the ID only exists after an insertion in the database.

Good point. So what should be the DX here?

**Scenario 1** - ID should not be null 
`.validate()` will fail for new entities (it shouldn't) and it will also fail for others UCs like update for entities with no IDs (it's ok)

**Scenario 2** - ID can be null 
`.validate()` will not fail for new entities (it's ok) but it will also not fail for others UCs like update for entities with no IDs (it shouldn't)

So, should we have many validators (ex: `.validateForUpdate()`)? Any ideas?

**Comment by thgpdr** [comment]
I dont know about many validators like```.validateForUpdate()```.
Maybe the responsibility of checking the id presence for a create/update scenario should be more manual, because its very specific.

What about ```.validate({ checkId: true })```?

**Comment by dalssoft** [comment]
> What about .validate({ checkId: true })?

Suggestion: `.validate({ exceptIDs: true })` - exceptions should be a opt-in

@thgpdr can you try implement it on this PR?

**Comment by thgpdr** [comment]
@dalssoft 
> Suggestion: .validate({ exceptIDs: true }) - exceptions should be a opt-in

nice, I like this one better! What about other fields? 
Should we give them the exception option too?


> @thgpdr can you try implement it on this PR?

sure, no problem!

**Comment by jhomarolo** [comment]
@thgpdr maybe like this:

`.validate().except([ 'id', 'createdAt', 'myOtherProperty']) `
or
`.validate( except : [ 'id', 'createdAt', 'myOtherProperty'] )`

what do you think?

**Comment by dalssoft** [comment]
@jhomarolo the reason I like `.validate({ exceptIDs: true })` more is because it has a clear use case: `Create XYZ` use cases. The CLI just need to create the file once and don't have to change it even if the entity changes.

`.validate( except : [ 'id', 'createdAt', 'myOtherProperty'] )` might be more flexible, but less declarative. However, It might be useful in the future for more advanced scenarios. For now I would go for `exceptIDs`

**Comment by jhomarolo** [comment]
@thgpdr any news about this PR?

**Pull request closed without merging**